descriptor/qsarsubstance/src/main/java/org/openscience/cdk/qsar/descriptors/substance/OxygenAtomCountDescriptor.java

```diff
@@ -0,0 +1,120 @@
+/* Copyright (C) 2011-2015  Egon Willighagen <egonw@users.sf.net>
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+package org.openscience.cdk.qsar.descriptors.substance;
+
+import org.openscience.cdk.exception.CDKException;
+import org.openscience.cdk.interfaces.IChemObjectBuilder;
+import org.openscience.cdk.interfaces.IIsotope;
+import org.openscience.cdk.interfaces.IMolecularFormula;
+import org.openscience.cdk.interfaces.ISubstance;
+import org.openscience.cdk.qsar.DescriptorSpecification;
+import org.openscience.cdk.qsar.DescriptorValue;
+import org.openscience.cdk.qsar.result.IDescriptorResult;
+import org.openscience.cdk.qsar.result.IntegerResult;
+import org.openscience.cdk.qsar.result.IntegerResultType;
+
+/**
+ * Descriptor that returns the number of oxygens in the chemical
+ * formula. Originally aimed at metal oxide nanoparticles.
+ *
+ * @author      egonw
+ * @cdk.githash
+ */
+public class OxygenAtomCountDescriptor implements ISubstanceDescriptor {
+
+	/** {@inheritDoc} */ @Override
+	public String[] getDescriptorNames() {
+        return new String[]{"NoMe"};
+	}
+
+	/** {@inheritDoc} */ @Override
+	public String[] getParameterNames() {
+        return new String[0];
+	}
+
+	/** {@inheritDoc} */ @Override
+	public Object getParameterType(String substance) {
+		return null;
+	}
+
+	/** {@inheritDoc} */ @Override
+	public Object[] getParameters() {
+		return new Object[0];
+	}
+
+	/** {@inheritDoc} */ @Override
+	public DescriptorSpecification getSpecification() {
+	    return new DescriptorSpecification(
+	        "http://egonw.github.com/resource/NM_001002",
+	        this.getClass().getName(),
+	        "The Chemistry Development Kit"
+	    );
+	}
+
+	/** {@inheritDoc} */ @Override
+	public void setParameters(Object[] parameters) throws CDKException {
+		return; // no parameters
+	}
+
+	/** {@inheritDoc} */ @Override
+	public DescriptorValue calculate(ISubstance substance) {
+        if (substance == null) return newNaNDescriptor();
+
+	    IMolecularFormula molFormula = SubstanceManipulator.getChemicalComposition(substance);
```

Comment:
Not intending to argue against your points, but I am trying to align things with the current design that:

- to ensure that the result matrix is rectangular and without gaps, fills with NaNs for descriptors for which something went wrong
- "exception" message are complementary; no current descriptor throws an Exception when something is wrong, because that messes up calculating a matrix
- some introspection is needed anyway, as a descriptor returns various result types already (per design). That can be a boolean, double, integer, as well as the array variants of multiple values are calculated (for many descriptor algorithms)

Add to that the observation that the code used by existing descriptors to return an new Integer((int)Double.NaN) which results in a '0', it seems that this patch uncovers some things in the QSAR descriptor API we do no longer like.

How shall we proceed? Should I introduce inconsistencies now, or fix these things in a big descriptor clean up round?

For example, I could update the code to return null values for the IDescriptorValue's returned (instead of whatever it now does when something goes wrong). That's an easy enough patch to run on all current code... but still at least a few hours work, I reckon... and a functional change, but we're not stable yet, so that should be OK.

Keep talking, please!